### ISSUE TITLE
Remove `EventedFd`

### ISSUE THREAD

**Issue by Thomasdezeeuw** [tokio-rs/mio#1013]
Should be remove `EventedFd` and instead implement `Evented` on `RawFd` directly?

**Comment by Thomasdezeeuw** [comment]
In #1014 I've renamed `EventedFd` to `SourceFd`, which isn't a very good name.

**Comment by carllerche** [comment]
I opted to not impl it on `RawFd` because not all fds can work with mio.

I don't love the name either, but I don't have a great option off the top of my head.

Another option would be to conditionally define `register_fd`, reregister_fd`, and `deregister_fd` on `Registry` (for unix).

**Comment by Thomasdezeeuw** [comment]
> I opted to not impl it on `RawFd` because not all fds can work with mio.

I understand this and agree, but handling `RawFd`s is already unsafe as many system calls don't work with any file descriptor (e.g. `accept` doesn't work with a fd that is a file (on disk)).

> I don't love the name either, but I don't have a great option off the top of my head.

`RegisterableFd` (not a big fan of it)? I don't really have a better option.

> Another option would be to conditionally define `register_fd`, `reregister_fd`, and `deregister_fd` on `Registry` (for unix).

I would prefer not to add these kind of methods as this could like to special methods for different OSes, e.g. `register_handle` on Windows etc.

**Comment by carllerche** [comment]
> I would prefer not to add these kind of methods as this could like to special methods for different OSes, e.g.

True, but I kind of am assuming that per-OS methods is a given at some level.

> `RegisterableFd` (not a big fan of it)? I don't really have a better option.

Can go w/ that if you wish.

**Comment by Thomasdezeeuw** [comment]
I do prefer a type such as `EventedFd` over adding three methods per OS. But I'll think about the name a bit more as I don't really like `RegisterableFd`.

**Comment by Thomasdezeeuw** [comment]
Had some more thoughts about this, creation and usage of a wrapper like `EventedFd` should probably be `unsafe` as the caller needs to ensure that file descriptor can actually be registered with epoll/kqueue. And my second though is that I would a matching API on the Windows side if possible, so this will likely need to wait until the rewrite is complete.

**Comment by carllerche** [comment]
I don't think it is `unsafe`... `unsafe` should be limited to APIs that could be unsound as per Rust's definition. Worst case scenario with passing in an invalid FD is you get an error.

**Comment by Thomasdezeeuw** [comment]
I've also used `unsafe` as a way to indicate caller requirements, in this case that would be ensuring the fd can be used in the system calls.

**Comment by carllerche** [comment]
I don't think it should be used like that IMO. That would just be a regular bug. `unsafe` should be reserved for cases that relate to APIs that incorrect usage would result in unsoundness as defined by the Rust memory model.

**Comment by Thomasdezeeuw** [comment]
An example of an `unsafe` function that wouldn't be unsound, in regards to memory the memory model, would be [`FromRawFd::from_raw_fd`](https://doc.rust-lang.org/std/os/unix/io/trait.FromRawFd.html#tymethod.from_raw_fd). If it would be used with an invalid/incorrect fd most system calls on it (e.g. when used with `File` and it's `io::Read` impl) would also return errors, but wouldn't cause memory unsoundness.

But I'm ok with keeping it as is, but I would really like a better name then `SourceFd`.
